src/engine/ResultTable.h

```diff
@@ -21,6 +25,59 @@ using std::mutex;
 using std::unique_lock;
 using std::vector;
 
+// The local vocabulary for a particular result table. It maps the IDs that are
+// not part of the normal vocabulary
+//
+//
+// It contains a map from
+// (local vocab) ids
+class LocalVocab {
+ public:
+  // Create a new, empty local vocabulary.
+  LocalVocab() {}
+
+  // Prevent accidental copying of a local vocabulary.
+  // TODO: Needed in SparqlExpressionTestHelpers.h:91.
+  // LocalVocab(const LocalVocab&) = delete;
+
+  // Get ID of a word in the local vocabulary. If the word was already
+  // contained, return the already existing ID. If the word was not yet
+  // contained, add it, and return the new ID.
+  [[maybe_unused]] Id getIdAndAddIfNotContained(const std::string& word);
+
+  // Start the construction of a local vocabulary. This is currently allowed
+  // only once, when the vocabulary is still empty.
+  void startConstructionPhase();
+
+  // Signal that the construction of the local vocabulary is done. This call
+  // will clear the `wordsToIdsMap_` (to save space) and afterwards,
+  // `getIdAndAddIfNotContained` can no longer be called.
+  void endConstructionPhase();
+
+  // The number of words in the vocabulary.
+  size_t size() const { return words_.size(); }
+
+  // Return true if and only if the local vocabulary is empty.
+  bool empty() const { return words_.empty(); }
+
+  // Return a const reference to the i-th word.
+  const std::string& operator[](size_t i) const { return words_[i]; }
```

Comment:
Shouldn't this interface be
`getWordFromId(Id)` or rather `getWord(LocalVocabIndex)`?
And probably these functions can afford the `at()` check for the size I think (local vocab is "expensive" anyway.

Reply:
I know what it is that bugged me.
Currently the `ResultTable` class has an `IdToOptionalString` method.
But this probably should removed, and the calling code should work on the `LocalVocab` directly which should then have an `IndexToOptionalString` method.